### ISSUE TITLE
[iOS] Fairplay with different key per asset [AVAssetResourceLoadingRequest finishLoading] was sent to an instance of AVAssetResourceLoadingRequest that was already finished. Ignoring.

### ISSUE THREAD

**Issue by daviCalavera** [TheWidlarzGroup/react-native-video#2881]
# Bug

## Platform
iOS (Tested on devices: iPhone 12 with iOS 16 & iPhone 8 with iOS 15)

## Environment info

React native info output:

```bash
System:
    OS: macOS 12.6
    CPU: (4) x64 Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz
    Memory: 1.31 GB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.17.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v16.17.1/bin/npm
    Watchman: 2022.09.26.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 16.0, macOS 12.3, tvOS 16.0, watchOS 9.0
    Android SDK:
      API Levels: 28, 29, 30, 31
      Build Tools: 28.0.3, 29.0.2, 29.0.3, 30.0.0, 30.0.2, 30.0.3, 31.0.0, 31.0.0, 31.0.0, 33.0.0, 33.0.0
      System Images: android-28 | Android TV Intel x86 Atom, android-29 | Google Play Intel x86 Atom, android-Q | Android TV Intel x86 Atom
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.3 AI-213.7172.25.2113.9014738
    Xcode: 14.0.1/14A400 - /usr/bin/xcodebuild
  Languages:
    Java: javac 19 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2
    react-native: 0.68.2 => 0.68.2
```

Library version: 5.2.1

## Steps To Reproduce

1. Open the app
2. select FairPlay content with single-drm key
3.  custom getLicense fetch the license and video starts to play.
4. select Fairplay content with multi-drm key (Key for AUDIO stream and a different key for VIDEO).
5. custom getLicense calls for each asset and gets the licenses correctly.
6. RCTVideo setLicense calls twice but log and error and the content doesn't start to play.

Error:
```
*** -[AVAssetResourceLoadingRequest finishLoading] was sent to an instance of AVAssetResourceLoadingRequest that was already finished. Ignoring.
```
...

## Expected behaviour

1. RCTVideo must set each key successfully
2. content must play on Player.

## Reproducible sample code

Unfortunately I cannot share the links of the content by contract and I didn't find a similar example content on the Web.
But it is a standard configuration recommended by the industry, to have assets encrypted with different keys.
A key for AUDIO assets and a another key for VIDEO assets.

**Comment by facugu1998** [comment]
Hello @daviCalavera. We are facing the same problem, did you find a solution?

**Comment by freeboub** [comment]
I need to create a PR for this, I think i found an easy solution for this issue !

**Comment by facugu1998** [comment]
Perfect. When do you estimate it will be ready? Is the solution available in any branch?

**Comment by freeboub** [comment]
Here is the patch I made on master (I didn't plan backport).
Honestly I am lost in documentation update ...

The issue is that onGetLicense is only called with drm.licenseUrl passed as a prop.
With this patch, if we don't pass license URL as prop, it will provide to the callback the license url provided in m3u.

This url is inform : skd://
Then, it is up to the application to reformat it in order to have a "real web url" 

Not sure on 5.2 it acts exactly in the same way, but this patch was mandatory for one of my customer integration.

I hope it helps (before a clean PR)
```

diff --git a/ios/Video/Features/RCTResourceLoaderDelegate.swift b/ios/Video/Features/RCTResourceLoaderDelegate.swift
index e1f1152..f13918e 100644
--- a/ios/Video/Features/RCTResourceLoaderDelegate.swift
+++ b/ios/Video/Features/RCTResourceLoaderDelegate.swift
@@ -134,8 +134,8 @@ class RCTResourceLoaderDelegate: NSObject, AVAssetResourceLoaderDelegate, URLSes
                 base64Certificate:_drm.base64Certificate
             ) .then{ spcData -> Void in
                 self._requestingCertificate = true
-                self._onGetLicense?(["licenseUrl": self._drm?.licenseServer ?? "",
-                                     "contentId": contentId,
+                self._onGetLicense?(["licenseUrl": self._drm?.licenseServer ?? loadingRequest.request.url?.absoluteString ?? "",
+                                     "contentId": contentId ?? "",
                                      "spcBase64": spcData.base64EncodedString(options: []),
                                      "target": self._reactTag])
             }

```

**Comment by facugu1998** [comment]
@freeboub We tested on master version with the fix you provided and it keeps not working. We debugged the app with xcode and realized that no requests to skd://{key}/AUDIO are being made. The only requests that it seems to be making are requesting license for the video skd://{key}/HD.

In the player, the first chunks are loaded and decrypted correctly, but there is no audio playing. And after some time, it stops playing even the video.

The difference with the 5.2.1 is that in that version the video is not playing at all, and it throws the exception detailed by the @daviCalavera issue. However it requests for both contents ids skd://{key}/HD and skd://{key}/AUDIO.

**Comment by freeboub** [comment]
@facugu1998 thanks for the detailed feedback! As I don't have such stream to test, I am afraid you will have to understand the root cause by yourself...
I cannot give more advise for now, sorry

**Comment by facugu1998** [comment]
@freeboub We have a public stream to test the problem with.
```
URL: https://cdnw1-stg.zetatv.com.uy/testzapp/bigbuckbunny/playlist.m3u8
licenseServer: https://pm-stg.zetatv.com.uy:8443/api/fairplay/get_license/fbdd541b-0036-4b4c-afef-351b7627ba81/
certificateUrl: https://pm-stg.zetatv.com.uy:8443/fairplay/cert.crt

import { Base64 } from "js-base64";
 getLicense: (spc, skd) => {
          return fetch(licenseServer, {
            method: "POST",
            headers: {
              "Content-Type": "application/octet-stream",
              keyId: `skd://${skd}`,
            },
            body: Base64.toUint8Array(spc),
          })
            .then((response) => response.blob())
            .then(async (response) => {
              const base_64 = await blobToBase64(response);
              return base_64.split(",")[1];
            })
            .catch((error) => {
              console.error("Error", error);
            });
        },
```


We would appreciate it if you could test it.

**Comment by facugu1998** [comment]
@freeboub Inspecting the native code we found a possible solution. We saw that `handleDrm` function in RCTResourceLoaderDelegate.swift sets `_requestingCertificate` the first time it runs. So, when its called with another key, for example, the one with "/AUDIO", it returns immediately without running getLicense.

The first solution we found assumes that the key with "/AUDIO" at the end is the last being called, so, conditioning the `_requestingCertificate` change to this call works properly in this case:

```
// RCTResourceLoaderDelegate.swift line 136
if (contentId?.range(of:"AUDIO") != nil){
    self._requestingCertificate = true
}
```

However this is no always true, so we decided to remove the following condition:

```
// RCTResourceLoaderDelegate.swift line 116
if _requestingCertificate {
            return true
}
 ```

In this case, for every different key, the function `handleDrm` runs and requests the corresponding license. You can test it with the stream we previously provided.


In order to get this working, we first got to return the full contentId to "onGetLicense" function, changing the following line:

```
let contentId = _drm.contentId ?? loadingRequest.request.url?.host
```

To

```
let contentId = _drm.contentId ?? loadingRequest.request.url?.absoluteString
```

**Comment by facugu1998** [comment]
Hello @freeboub, were you able to check out our proposed solution?

**Comment by freeboub** [comment]
linked PR has been merged in 6.0.0-alpha.8, let's close this issue
